src/main/java/seedu/address/logic/parser/FindCommandParser.java

```diff
@@ -20,15 +28,53 @@
      * @throws ParseException if the user input does not conform the expected format
      */
     public FindCommand parse(String args) throws ParseException {
-        String trimmedArgs = args.trim();
+        String trimmedArgs = " 0 " + args.trim();
         if (trimmedArgs.isEmpty()) {
             throw new ParseException(
                     String.format(MESSAGE_INVALID_COMMAND_FORMAT, FindCommand.MESSAGE_USAGE));
         }
 
-        String[] nameKeywords = trimmedArgs.split("\\s+");
+        ArgumentMultimap argMultimap =
+                ArgumentTokenizer.tokenize(trimmedArgs, PREFIX_TITLE, PREFIX_SERIAL_NUMBER, PREFIX_AUTHOR,
+                        PREFIX_GENRE);
 
-        return new FindCommand(new TitleContainsKeywordPredicate(Arrays.asList(nameKeywords)));
+        BookPredicate predicate = new BookPredicate();
+
+        if (argMultimap.getValue(PREFIX_TITLE).isPresent()) {
+            predicate.addTitle(ParserUtil.parseTitle(argMultimap.getValue(PREFIX_TITLE).get()).value);
+        }
+
+        if (argMultimap.getValue(PREFIX_SERIAL_NUMBER).isPresent()) {
+            predicate.addSerialNumber(ParserUtil.parseSerialNumber(argMultimap.getValue(PREFIX_SERIAL_NUMBER)
+                    .get()).value);
+        }
+
+        if (argMultimap.getValue(PREFIX_AUTHOR).isPresent()) {
+            predicate.addAuthor(ParserUtil.parseAuthor(argMultimap.getValue(PREFIX_AUTHOR).get()).value);
+        }
+
+        parseGenresForFind(argMultimap.getAllValues(PREFIX_GENRE)).ifPresent(predicate::addGenres);
+
+        if (!predicate.isValid()) {
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, FindCommand.MESSAGE_USAGE));
+        }
+
+        return new FindCommand(predicate);
+    }
+
+    /**
+     * Parses {@code Collection<String> genres} into a {@code Set<Genre>} if {@code genres} is non-empty.
+     * If {@code genres} contain only one element which is an empty string, it will be parsed into a
+     * {@code Set<Genre>} containing zero genres.
+     */
+    private Optional<Set<Genre>> parseGenresForFind(Collection<String> genres) throws ParseException {
+        assert genres != null;
```

Comment:
Would you consider using `requireNonNull` here instead?

Reply:
yep! nice catch!